server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SLOTS.java

```diff
@@ -89,32 +88,34 @@ private static CompletionStage<CharSequence> getSlotsInformation(Resp3Handler ha
       return requestNodesNetworkInformation(hash.getMembers(), handler)
             .thenApply(information -> {
                StringBuilder builder = new StringBuilder();
-
-               Map<List<Address>, IntSet> segmentOwners = new HashMap<>();
-               for (int i = 0; i < hash.getNumSegments(); i++) {
-                  segmentOwners.computeIfAbsent(hash.locateOwnersForSegment(i), ignore -> IntSets.mutableEmptySet())
-                        .add(i);
-               }
-
                int size = 0;
-               for (Map.Entry<List<Address>, IntSet> entry : segmentOwners.entrySet()) {
-                  List<Address> owners = entry.getKey();
-                  IntSet segments = entry.getValue();
-
-                  for (int i = segments.nextSetBit(0); i >= 0; i = segments.nextSetBit(i + 1)) {
-                     int runStart = i;
-                     while (segments.contains(i + 1)) {
-                        i++;
-                     }
 
-                     size++;
-                     builder.append('*').append(2 + owners.size()).append(CRLF_STRING);
-                     builder.append(':').append(runStart).append(CRLF_STRING);
-                     builder.append(':').append(i).append(CRLF_STRING);
-                     for (Address owner: owners) {
-                        String info = information.get(owner);
-                        builder.append(info);
+               SegmentSlotRelation ssr = handler.respServer().segmentSlotRelation();
+
+               // These two will always be initialized in first loop below
+               List<Address> ownersForSegment = null;
+               IntSet slots = IntSets.mutableEmptySet();
+
+               int totalSegmentCount = hash.getNumSegments();
+               for (int i = 0; i < totalSegmentCount; ++i) {
+                  List<Address> currentOwners = hash.locateOwnersForSegment(i);
+                  slots.addAll(ssr.segmentToSlots(i));
```

Comment:
Even better here would be just to keep the start and end Segment and then when printing just use the slotWidth like the SHARDS command is doing.

Reply:
Updated.